R/xgx_stat_ci.R

```diff
@@ -99,87 +99,150 @@
+      if (is.null(lay$aes_params$size)) lay$aes_params$size <- 2
     }
     else if (igeom == "line") {
-      if (is.null(temp$aes_params$size)) temp$aes_params$size <- 1
+      if (is.null(lay$aes_params$size)) lay$aes_params$size <- 1
     }
     else if (igeom == "errorbar") {
-      if (is.null(temp$aes_params$size)) temp$aes_params$size <- 1
-      if (is.null(temp$geom_params$width)) {
-        temp$geom_params$width <- 0
-      }
+      if (is.null(lay$aes_params$size)) lay$aes_params$size <- 1
+      if (is.null(lay$geom_params$width)) lay$geom_params$width <- 0
     }
     else if (igeom == "ribbon") {
-      if(is.null(temp$aes_params$alpha)) temp$aes_params$alpha <- 0.25
+      if(is.null(lay$aes_params$alpha)) lay$aes_params$alpha <- 0.25
     }
     else if (igeom == "pointrange") {
-      if(is.null(temp$aes_params$size)) temp$aes_params$size <- 1
-      temp$geom$geom_params$fatten <- 2
+      if(is.null(lay$aes_params$size)){
+        lay$aes_params$size <- 1
+        lay$geom$geom_params$fatten <- 2
+      }
     }
-    
 
-    ret[[paste0("geom_", igeom)]] <- temp
+    lays[[paste0("geom_", igeom)]] <- lay  
   }
-
-  return(ret)
+  
+  return(lays)
 }
+
+
+#' Stat ggproto object for creating ggplot layers of binned confidence intervals
+#'
+#' 
+#' \code{StatCI} returns a ggproto object for plotting mean +/- confidence bins
+#' 
+#'
+#' @return ggplot2 ggproto object
+#'
+#' @export
+StatCI <- ggplot2::ggproto("StatCI", ggplot2::Stat,
+                  
+    required_aes = c("x","y"),
+    
+    compute_group = function(data, scales, conf_level, distribution, bins) {
+    
+      if (distribution == "ordinal") {
+        stat_data <- data
+      }
+      else {
+        stat_data <- data %>% mutate(quantile_index = dplyr::ntile(data$x, bins)) %>%
+                              group_by(quantile_index) %>%
+                              summarize(x = median(x),
+                                        ymin = xgx_conf_int(y, conf_level, distribution)$ymin,
+                                        ymax = xgx_conf_int(y, conf_level, distribution)$ymax,
+                                        y = xgx_conf_int(y, conf_level, distribution)$y) %>%
+                              ungroup()
+      }
+
+      return(stat_data)
+    },
+    
+    setup_data = function(data, params){
+      
+      # If the data is ordinal, calculate percentages
+      #   for each category across each bin
+      if (params$distribution == "ordinal"){
+        
+        # Get median x value for each bin
+        median_x <- data %>% mutate(quantile_index = dplyr::ntile(data$x, params$bins)) %>%
+                              group_by(quantile_index) %>%
+                              summarize(x = median(x))
+        
+        # Get the number of each category in each bin 
+        counts <- data %>% mutate(quantile_index = dplyr::ntile(data$x, params$bins)) %>%
+                              group_by(quantile_index, y) %>%
```

Comment:
do we need group_by(quantile_index, y, _colour, fill, group, PANEL_)? Then we can remove these from the summarize lines 230 through 233